hazelcast/src/main/java/com/hazelcast/sql/impl/QueryId.java

```diff
@@ -58,22 +58,72 @@ public QueryId() {
      * @return Query ID.
      */
     public static QueryId create(UUID memberId) {
-        UUID qryId = UuidUtil.newUnsecureUUID();
+        UUID localId = UuidUtil.newUnsecureUUID();
 
         return new QueryId(
             memberId.getMostSignificantBits(),
             memberId.getLeastSignificantBits(),
-            qryId.getMostSignificantBits(),
-            qryId.getLeastSignificantBits()
+            localId.getMostSignificantBits(),
+            localId.getLeastSignificantBits()
         );
     }
 
+    public static QueryId parse(String input) {
+        assert input != null;
+
+        String[] parts = input.split("_");
+
+        if (parts.length != 2) {
+            throw new IllegalArgumentException("Query ID is malformed: " + input);
+        }
+
+        try {
+            UUID memberId = UUID.fromString(parts[0]);
+            UUID localId = UUID.fromString(parts[1]);
```

Comment:
Javadoc of `UUID.toString()` specifies that the result is always 36 chars, so we could just split there. But `UUID.fromString` doesn't assume that, so I don't know. The below code is safe in this regard and doesn't deal with regexes.

```suggestion
        int underscorePos = input.indexOf("_");

        if (underscorePos < 0) {
            throw new IllegalArgumentException("Query ID is malformed: " + input);
        }

        try {
            UUID memberId = UUID.fromString(input.substring(0, underscorePos));
            UUID localId = UUID.fromString(input.substring(underscorePos + 1);
```